[Ansor][AutoTVM v2.0] Phase 1: Add cache_read/cache_write steps

jcf94 opened the pull request:
For the full upstream plan, see [Ansor RFC](https://discuss.tvm.ai/t/rfc-ansor-an-auto-scheduler-for-tvm-autotvm-v2-0/7005/32).

In this PR, we bring cache_read/cache_write steps for Ansor auto_scheduler.

These steps will insert extra stage to the original ComputeDAG, the class member `current_compute_dag` in State is used to track the up-to-date ComputeDAG.

cc @merrymercy @comaniac @junrushao1994 @FrozenGene

comaniac commented on src/auto_scheduler/loop_state.h:
```diff
@@ -195,6 +197,17 @@ class AttachMap : public ObjectRef {
                    const std::vector<IterKey>& new_iters);
 
+  /*!
+   * \brief Traverse through `stage_to_attach_iter` and `iter_to_attached_stages` map, add offset
+   * to stage indexes that are larger than the start_id. Used for steps that inserts net stages to
```
You mean "and update offset"?

comaniac commented on src/auto_scheduler/loop_state.h:
```diff
@@ -195,6 +197,17 @@ class AttachMap : public ObjectRef {
+   * \param offset The index offset to be added to the stage index.
+   * \return The updated AttachMap after applying stage index offset.
+   */
+  AttachMap ApplyStageIdOfffset(int start_id, int offset) const;
```
s/ApplyStageIdOfffset/ApplyStageIdOffset/

jcf94 commented on python/tvm/auto_scheduler/loop_state.py:
```diff
@@ -371,6 +433,23 @@ def _update_stage_id_map(self):
+        self.stage_id_map[added_op] = new_stage_id
+
+        # Update stage_id_map for new ops
+        self._update_stage_id_map()
```
After cache_read/cache_write, actually all the stages behind the new added stage will be update, their ops are different.
This is to keep the original Tensor/ops works and add those new ops to the id map.

comaniac replied:
So we can imagine the stage ID map will have all ops, including out-of-date and up-to-date ops. Whatever the op is up-to-date, this function makes sure it points to the right stage.

If the above statement about my understanding to the stage ID map is correct, then I agree with you that this is the must. Meanwhile, we need to add more explanations.

jcf94 commented on python/tvm/auto_scheduler/loop_state.py:
```diff
@@ -371,6 +433,23 @@ def _update_stage_id_map(self):
+    def _insert_new_stage(self, new_stage_id):
+        added_op = self.stages[new_stage_id].op
+
+        # Add a new stage will change all ops. But we still want to use the old ops to index stages,
+        # So we keep updating them and do not remove the old ops.
```
- This is an internal function which should not be called outside, seems the docstring is unnecessary.
- This function actuall means insert a new stage to the id map(add should modify the stage id behind this stage after the insert), will consider for a better name.
- Good point.
- This will be used in Rfactor, too.

jcf94 replied:
Update to a `_apply_stage_id_offset`, this plays a similar role as the AttachMap::ApplyStageIdOffset.

comaniac replied:
- I know we don't need docstring for internal functions, but it's not a problem if we need. At least we should put the description in the beginning of the function.
- `_apply_stage_id_offset` sounds better.
- I didn't mean this function will be used by CacheRead/Write only. I meant this function is required when we implement a step that changes the stages, but there's no checking to enforce that. An ideal solution would be having a function to deal with stage changing and let CacheRead/CacheWrite/Rfactor call the function.

jcf94 commented on python/tvm/auto_scheduler/loop_state.py:
```diff
@@ -351,6 +351,68 @@ def compute_root(self, stage):
+        -----
+        Cache write step will add an extra stage to the original ComputeDAG, a up-to-date
+        ComputeDAG is stored in State's `current_compute_dag`.
+        This step will cache write all output tensors of the target stage.
```
The `current_compute_dag` should be a class member only used in C++. Removed this from the python doc.

jcf94 commented on python/tvm/auto_scheduler/loop_state.py:
```diff
@@ -351,6 +351,68 @@ def compute_root(self, stage):
+
+        Notes
+        -----
+        Cache read step will add an extra stage to the original ComputeDAG.
```
Updated.

tqchen commented:
cc @jroesch @jwfromm

comaniac commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -923,5 +958,275 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
 
+/********** Primitives adding new stages **********/
+
+// Common part for steps that add new stages
+// (e.g. CacheReadStep, CacheWriteStep, RfactorStep)
+void AddStageModificationSteps(int step_id, const Array<Step>& transform_steps,
+                               Array<Step>* replay_steps) {
```
The description and function name could be improved. This function checks if `step_id` is a step that adds a new stage, and puts the step to `replay_steps` if so. Will this function be extended when adding Rfactor? If not, then it seems to me that this function is unnecessary.

jcf94 replied:
Yes, and the Rfactor will involve more complicated check. I have a TODO hint behind.

jcf94 replied:
Moved the step traverse parts to this fucntion, updated the function name.

comaniac commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -923,5 +958,275 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+String CacheReadStepNode::PrintAsPythonAPI(Array<te::Stage>* stages, StageToAxesMap* stage_to_axes,
+                                           te::Schedule* schedule) const {
+  std::stringstream ss;
+  // Copy stage here, for the original stage will change after apply
```
Improve this comment.

comaniac commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -923,5 +958,275 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+  int added_ops = current_compute_dag->ops.size() - last_dag_op_size;
+  CHECK_GE(added_ops, 1);
+
+  // target -> target_compute + target
+  // Assume target stage has never been applied any steps before cache_write
+  // Should insert new stage, update target stage, update the later stage's op
```
```suggestion
  // target_stage -> cache_write_stage + target_stage
  // Assume no step has been applied to the target stage cache_write.
  // Insert a new cache write stage, update the target stage and the op-to-stage-id map for ops in later stages.
```

comaniac commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -923,5 +958,275 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+                        Stage(current_compute_dag->ops[stage_id]));
+  pstate->stages.Set(stage_id + 1, Stage(current_compute_dag->ops[stage_id + 1]));
+  int next_stage_id = stage_id + 2;
+  // Notice: added_ops should actually assert to be 1
+  // branch of 2 here is somehow a hack to TVM's cache_write bug with multi outputs
+  // see `tests/python/unittest/test_auto_scheduler_loop_state.py::test_cache_read_write` test for
+  // more information
+  // TODO(jcf94): Fix the cache write bug in TVM and remove these branches here
```
```suggestion
  // TODO(jc94): Fix the cache write bug in TVM and remove added_op == 2 support.
  // TVM's cache_write has a bug with multi outputs.
  // See `tests/python/unittest/test_auto_scheduler_loop_state.py::test_cache_read_write` test.
```

comaniac commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -923,5 +958,275 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+                             : dag->ops.size();
+  const ComputeDAG& current_compute_dag = dag.ReplayAndGetDAG(replay_steps);
+  int added_ops = current_compute_dag->ops.size() - last_dag_op_size;
+  CHECK_GE(added_ops, 1);
```
According to L1150, please add a TODO here as well to change to `CHECK_EQ`.

merrymercy commented on python/tvm/auto_scheduler/loop_state.py:
```diff
@@ -351,6 +351,72 @@ def compute_root(self, stage):
 
+    def cache_read(self, stage, scope_name, reader_stages):
+        """ Schedule primitive corresponds to te.schedule.cache_read.
+
```
This step does the same thing as `te.schedule.cache_read` does. We choose to not refer to `te.schedule.cache_read` instead of copying the doc string from it.

merrymercy replied:
Maybe we can make the pointer more clear. (e.g., say "see also `te.schedule.cache_read`")

jcf94 replied:
Add a pointer to `te.schedule.cache_read`, we may also add this to other steps later.

jcf94 commented on tests/python/unittest/test_auto_scheduler_loop_state.py:
```diff
@@ -143,6 +143,282 @@ def test_compute_at_root_inline():
+    #
+    # Seems there's bug with the input/output tensor. Such multi outputs case
+    # should be unusual, so we make some hack on DoCacheWrite
+    # To be fixed in the future
```
I think these TODOs are necessary, or others may be confused seeing these codes.
We dose have a track issue for all those needs to be handled: https://github.com/merrymercy/Ansor/issues/65

merrymercy replied:
This is a bug of tvm and is outside the scope of Ansor. It is never used in Ansor either. We just found this bug by accident during development. So we can just skip it

jcf94 commented on src/auto_scheduler/loop_state.h:
```diff
@@ -195,6 +197,17 @@ class AttachMap : public ObjectRef {
 
+  /*!
+   * \brief Traverse through `stage_to_attach_iter` and `iter_to_attached_stages` map, add offset
+   * to stage indexes that are larger than the start_id. Used for steps that inserts net stages to
```
Oh, this seems a typo ... should be `new stages`.

merrymercy commented on src/auto_scheduler/loop_state.h:
```diff
@@ -225,6 +238,13 @@ class StateNode : public Object {
    */
   AttachMap attach_map;
+  /*!
+   * \brief The up-to-date ComputeDAG of this state, used for some steps that may change the
```
This state is always up-to-date.
The comment in the above method says the "initial dag" may not be up-to-date. So we need to store a up-to-date dag here.

merrymercy commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -923,5 +958,272 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+
+/*!
+ * \brief Common part for steps that add new stages(e.g. CacheReadStep, CacheWriteStep,
+ * RfactorStep). This will filter out all steps that can change the stages of ComputeDAG.
```
```suggestion
 * RfactorStep). This will filter out all steps that can change the number of stages in a ComputeDAG.
```

merrymercy commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -923,5 +958,272 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+  const ComputeDAG& current_compute_dag =
+      dag.ReplayAndGetDAG(GetStageModifiableSteps(GetRef<Step>(this), (*state)->transform_steps));
+  int added_ops = current_compute_dag->ops.size() - last_dag_op_size;
+  // TODO(jcf94): Update this check to equal after fixing the cache write bug in TVM
```
Yeah, I can open an issue later.
For this specific line, this is a bug of tvm and this feature is not used in Ansor. So it is okay to skip this.

merrymercy commented on src/auto_scheduler/transform_step.h:
```diff
@@ -659,6 +671,153 @@ class ComputeRootStep : public Step {
+
+/*!
+ * \brief Cache read step that corresponds to te::Schedule::cache_read.
+ * \note Cache read step will add an extra stage to the original ComputeDAG, a up-to-date ComputeDAG
```
The `current_compute_dag` in the State is always up-to-date.
The `compute_dag` in SearchTask may not be up-to-date

jcf94 replied:
Updated the explaination.

merrymercy commented on src/auto_scheduler/transform_step.h:
```diff
@@ -659,6 +671,153 @@ class ComputeRootStep : public Step {
+ */
+class CacheReadStepNode : public StepNode {
+ public:
+  /*! \brief The scope name to be set for the new added read stage. (e.g. local, shared, global) */
```
```suggestion
  /*! \brief The scope name to of the newly added read stage. (e.g. local, shared, global) */
```

merrymercy commented on src/auto_scheduler/transform_step.h:
```diff
@@ -659,6 +671,153 @@ class ComputeRootStep : public Step {
+class CacheWriteStepNode : public StepNode {
+ public:
+  /*!
+   * \brief The scope name to be set for the new added write stage. (e.g. local, shared,
```
ditto

jcf94 commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -923,5 +958,272 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+      ret_steps.push_back(step);
+    }
+    // TODO(jcf94): add rfactor support
+    if (step.same_as(current_step)) {
```
Updated.

jcf94 commented:
@merrymercy @junrushao1994 @jroesch @comaniac Comments are addressed, please take another look.
I'll submit the PRs for the rest steps soon after this.

merrymercy commented:
I opened an issue to track the unresolved todo items during upstream.
However, all todos in this PR are not related to Ansor and are outside the scope of Ansor.
They are just minor bugs of tvm we found during the development. We don't have the plan to fix them because we don't use these features.

merrymercy commented on python/tvm/auto_scheduler/compute_dag.py:
```diff
@@ -126,11 +126,16 @@ def infer_bound_from_state(self, state):
         state_obj = state if isinstance(state, StateObject) else state.state_object
-        return State(_ffi_api.ComputeDAGInferBoundFromState(self, state_obj), self)
+        updated_state = State(_ffi_api.ComputeDAGInferBoundFromState(self, state_obj), self)
+        # Copy the stage_id_map from the original state
```
```suggestion
        # Copy the stage_id_map from the original state to make the old indices still valid.
```

merrymercy commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -923,5 +958,275 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+
+/*!
+ * \brief Common part for steps that add new stages(e.g. CacheReadStep, CacheWriteStep,
+ * RfactorStep). This will filter out all steps that can change the number of stages in a
```
```suggestion
 * RfactorStep). This returns all steps that can change the number of stages in a
```
"Filter out" is the antonym of "return"

merrymercy commented on src/auto_scheduler/loop_state.h:
```diff
@@ -225,6 +238,16 @@ class StateNode : public Object {
    */
   AttachMap attach_map;
+  /*!
+   * \brief The up-to-date ComputeDAG of this state, used for some steps that may change the
+   * stage structure of the ComputeDAG (e.g. CacheReadStep/CacheWriteStep). This will alway be kept
+   * up-to-date, while the original ComputeDAG may not be up-to-date.
+   * The default value is an empty NullOpt, means no modification to the original DAG.
+   * Typical usage for this is when acquiring information from ComputeDAG (e.g. check for its
+   * AccessAnalyzer), use the `current_compute_dag` first, if it's Null, use the original
+   * ComputeDAG.
+   */
```
```suggestion
   * \brief The up-to-date ComputeDAG of this state. If the value is null, it means no steps modify the DAG. 
   * The dag is the same as the compute_dag in the `SearchTask`. Otherwise, it means some steps 
   * (e.g., CacheReadStep/CacheWriteStep) modify the dag, then the stored value is the new and up-to-date
   * DAG for this state.
   */
```

merrymercy commented on src/auto_scheduler/transform_step.h:
```diff
@@ -194,20 +201,25 @@ void StepApplyToState(const Step& step, State* state, const ComputeDAG& dag);
  * \param step The step to be applied to tvm.schedule.
- * \param stages A pointer to a `te::Stage` Array.
- * \param stage_to_axes A pointer to a StageToAxesMap.
+ * \param stages A mutable pointer to a `te::Stage` Array.
```
This kind of doc only repeats the type of the parameters and gives no information

jcf94 replied:
Updated.

zhiics commented on src/auto_scheduler/transform_step.cc:
```diff
@@ -923,5 +958,275 @@ String ComputeRootStepNode::PrintAsPythonAPI(Array<te::Stage>* stages,
+ * RfactorStep). This will return all steps that can change the number of stages in a ComputeDAG,
+ * and stop by the current step.
+ */
+Array<Step> GetFormerStageModifiableSteps(Step current_step, const Array<Step>& transform_steps) {
```
This naming seems not quite clear. How about `GetMutableSteps`?

jcf94 replied:
Or `StageMutableSteps` ?
It seems that just `MutableSteps` is still not quite clear.

merrymercy commented:
@jcf94  Sorry, you have to rebase because I moved hearder files to the public folder `include/tvm/auto_scheduler`

jcf94 commented:
> @jcf94 Sorry, you have to rebase because I moved hearder files to the public folder `include/tvm/auto_scheduler` in #6103

Fine, updated with the new upstream master.

merrymercy commented:
@jroesch @comaniac  please take another look and approve

merrymercy commented:
I dismissed @jroesch 's review because all of his comments are addressed.
Most of his concern is due to some todo items caused by a minor TVM bug, which is outside of the scope of this PR.
Follow his suggestion, I also opened an issue (#6133) to track other todos.
